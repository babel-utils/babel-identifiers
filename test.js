// @flow
'use strict';

const pluginTester = require('babel-plugin-tester');
const printAST = require('ast-pretty-print');
const getBabylonOptions = require('babylon-options');
const {getIdentifierKind} = require('./');

const babelOptions = {
  parserOpts: getBabylonOptions({
    stage: 0,
    plugins: ['flow'],
  }),
};

const print = (t, value) => {
  let printed = printAST(value) + '\n';
  return t.expressionStatement(
    t.templateLiteral([
      t.templateElement({ raw: printed }, true)
    ], [])
  );
};

const plugin = ({types: t}) => {
  return {
    name: 'test-plugin',
    visitor: {
      Program(path) {
        let ids = [];

        path.traverse({
          'Identifier|TypeParameter'(path) {
            ids.push({
              name: path.node.name,
              kind: getIdentifierKind(path),
            });
          },
        });

        path.pushContainer('body', print(t, ids));
      },
    },
  };
};

pluginTester({
  plugin,
  snapshot: true,
  babelOptions,
  tests: {
    'member expression': 'a.b',
    'member expression computed': 'a[b]',
    'object property': '({ a: b })',
    'object property computed': '({ [a]: b })',
    'object method': '({ a() {} })',
    'object method computed': '({ [a]() {} })',
    'let object pattern': 'let {a} = {}',
    'let object pattern nested': 'let {a: {b}} = {}',
    'let object pattern renamed': 'let {a: b} = {}',
    'function param object pattern': '(({a}) => {})',
    'function param object pattern nested': '(({a: {b}}) => {})',
    'function param object pattern renamed': '(({a: b}) => {})',
    'array expression': '[a]',
    'array pattern': '[a] = []',
    'array pattern nested': '[[a]] = []',
    'let array pattern': 'let [a, ...b] = []',
    'let array pattern nested': 'let [[a, ...b]] = []',
    'let array pattern nested object': 'let { a: [b, ...c] } = {}',
    'let': 'let a = b',
    'function declaration': 'function a(b, ...c) {}',
    'function expression': '(function a(b, ...c) {})',
    'class declaration': 'class a {}',
    'class expression': '(class a {})',
    'for loop': 'for (let a;;) {}',
    'for loop reference': 'for (a;;) {}',
    'import default': 'import a from ""',
    'import named': 'import {a} from ""',
    'import named renamed': 'import {a as b} from ""',
    'export default from': 'export a from ""',
    'export namespace from': 'export * as a from ""',
    'export named': 'export {a}',
    'export named renamed': 'export {a as b}',
    'export named from': 'export {a} from ""',
    'export named from renamed': 'export {a as b} from ""',
    'type alias': 'type a = {}',
    'interface declaration': 'interface a {}',
    'type params declaration': 'type a<b: c = d> = {}',
    'class type param': 'class a<b> {}',
    'class extends': 'class a extends b {}',
    'class extends type param': 'class a extends b<c> {}',
    'class implements': 'class a implements b {}',
    'class implements type param': 'class a implements b<c> {}',
    'function type return': 'type a = () => b',
    'function type param': 'type a = (b: c) => {}',
    'function type param unnamed': 'type a = (b) => {}',
    'function type type param': 'type a = <b>() => {}',
    'object type property': 'type a = { b: c }',
    'object type method': 'type a = { b(): c }',
    'object type indexer': 'type a = { [b: c]: d }',
    'object type computed property': 'type a = { [b]: c }',
  },
});
