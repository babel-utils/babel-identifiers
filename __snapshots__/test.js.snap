// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. member expression 1`] = `
"
a.b

      ↓ ↓ ↓ ↓ ↓ ↓

a.b;
\`
  -
      kind: 'reference'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
\`;
"
`;

exports[`2. member expression computed 1`] = `
"
a[b]

      ↓ ↓ ↓ ↓ ↓ ↓

a[b];
\`
  -
      kind: 'reference'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
\`;
"
`;

exports[`3. object property 1`] = `
"
({ a: b })

      ↓ ↓ ↓ ↓ ↓ ↓

({ a: b });
\`
  -
      kind: 'static'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`4. object property computed 1`] = `
"
({ [a]: b })

      ↓ ↓ ↓ ↓ ↓ ↓

({ [a]: b });
\`
  -
      kind: 'reference'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`5. object method 1`] = `
"
({ a() {} })

      ↓ ↓ ↓ ↓ ↓ ↓

({ a() {} });
\`
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`6. object method computed 1`] = `
"
({ [a]() {} })

      ↓ ↓ ↓ ↓ ↓ ↓

({ [a]() {} });
\`
  -
      kind: 'reference'
      name: 'a'
\`;
"
`;

exports[`7. let object pattern 1`] = `
"
let {a} = {}

      ↓ ↓ ↓ ↓ ↓ ↓

let { a } = {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`8. let object pattern nested 1`] = `
"
let {a: {b}} = {}

      ↓ ↓ ↓ ↓ ↓ ↓

let { a: { b } } = {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`9. let object pattern renamed 1`] = `
"
let {a: b} = {}

      ↓ ↓ ↓ ↓ ↓ ↓

let { a: b } = {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`10. function param object pattern 1`] = `
"
(({a}) => {})

      ↓ ↓ ↓ ↓ ↓ ↓

({ a }) => {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`11. function param object pattern nested 1`] = `
"
(({a: {b}}) => {})

      ↓ ↓ ↓ ↓ ↓ ↓

({ a: { b } }) => {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`12. function param object pattern renamed 1`] = `
"
(({a: b}) => {})

      ↓ ↓ ↓ ↓ ↓ ↓

({ a: b }) => {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`13. array expression 1`] = `
"
[a]

      ↓ ↓ ↓ ↓ ↓ ↓

[a];
\`
  -
      kind: 'reference'
      name: 'a'
\`;
"
`;

exports[`14. array pattern 1`] = `
"
[a] = []

      ↓ ↓ ↓ ↓ ↓ ↓

[a] = [];
\`
  -
      kind: 'reference'
      name: 'a'
\`;
"
`;

exports[`15. array pattern nested 1`] = `
"
[[a]] = []

      ↓ ↓ ↓ ↓ ↓ ↓

[[a]] = [];
\`
  -
      kind: 'reference'
      name: 'a'
\`;
"
`;

exports[`16. let array pattern 1`] = `
"
let [a, ...b] = []

      ↓ ↓ ↓ ↓ ↓ ↓

let [a, ...b] = [];
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`17. let array pattern nested 1`] = `
"
let [[a, ...b]] = []

      ↓ ↓ ↓ ↓ ↓ ↓

let [[a, ...b]] = [];
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`18. let array pattern nested object 1`] = `
"
let { a: [b, ...c] } = {}

      ↓ ↓ ↓ ↓ ↓ ↓

let { a: [b, ...c] } = {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`19. let 1`] = `
"
let a = b

      ↓ ↓ ↓ ↓ ↓ ↓

let a = b;
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`20. function declaration 1`] = `
"
function a(b, ...c) {}

      ↓ ↓ ↓ ↓ ↓ ↓

function a(b, ...c) {}
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
  -
      kind: 'binding'
      name: 'c'
\`;
"
`;

exports[`21. function expression 1`] = `
"
(function a(b, ...c) {})

      ↓ ↓ ↓ ↓ ↓ ↓

(function a(b, ...c) {});
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
  -
      kind: 'binding'
      name: 'c'
\`;
"
`;

exports[`22. class declaration 1`] = `
"
class a {}

      ↓ ↓ ↓ ↓ ↓ ↓

class a {}
\`
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`23. class expression 1`] = `
"
(class a {})

      ↓ ↓ ↓ ↓ ↓ ↓

(class a {});
\`
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`24. for loop 1`] = `
"
for (let a;;) {}

      ↓ ↓ ↓ ↓ ↓ ↓

for (let a;;) {}
\`
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`25. for loop reference 1`] = `
"
for (a;;) {}

      ↓ ↓ ↓ ↓ ↓ ↓

for (a;;) {}
\`
  -
      kind: 'reference'
      name: 'a'
\`;
"
`;

exports[`26. import default 1`] = `
"
import a from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

import a from \\"\\";
\`
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`27. import named 1`] = `
"
import {a} from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

import { a } from \\"\\";
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`28. import named renamed 1`] = `
"
import {a as b} from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

import { a as b } from \\"\\";
\`
  -
      kind: 'binding'
      name: 'b'
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`29. export default from 1`] = `
"
export a from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

export a from \\"\\";
\`
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`30. export namespace from 1`] = `
"
export * as a from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

export * as a from \\"\\";
\`
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`31. export named 1`] = `
"
export {a}

      ↓ ↓ ↓ ↓ ↓ ↓

export { a };
\`
  -
      kind: 'reference'
      name: 'a'
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`32. export named renamed 1`] = `
"
export {a as b}

      ↓ ↓ ↓ ↓ ↓ ↓

export { a as b };
\`
  -
      kind: 'reference'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
\`;
"
`;

exports[`33. export named from 1`] = `
"
export {a} from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

export { a } from \\"\\";
\`
  -
      kind: 'static'
      name: 'a'
  -
      kind: 'static'
      name: 'a'
\`;
"
`;

exports[`34. export named from renamed 1`] = `
"
export {a as b} from \\"\\"

      ↓ ↓ ↓ ↓ ↓ ↓

export { a as b } from \\"\\";
\`
  -
      kind: 'static'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
\`;
"
`;

exports[`35. type alias 1`] = `
"
type a = {}

      ↓ ↓ ↓ ↓ ↓ ↓

type a = {};
\`
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`36. interface declaration 1`] = `
"
interface a {}

      ↓ ↓ ↓ ↓ ↓ ↓

interface a {}
\`
  -
      kind: 'binding'
      name: 'a'
\`;
"
`;

exports[`37. type params declaration 1`] = `
"
type a<b: c = d> = {}

      ↓ ↓ ↓ ↓ ↓ ↓

type a<b: c = d> = {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`38. class type param 1`] = `
"
class a<b> {}

      ↓ ↓ ↓ ↓ ↓ ↓

class a<b> {}
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`39. class extends 1`] = `
"
class a extends b {}

      ↓ ↓ ↓ ↓ ↓ ↓

class a extends b {}
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`40. class extends type param 1`] = `
"
class a extends b<c> {}

      ↓ ↓ ↓ ↓ ↓ ↓

class a extends b<c> {}
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`41. class implements 1`] = `
"
class a implements b {}

      ↓ ↓ ↓ ↓ ↓ ↓

class a implements b {}
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`42. class implements type param 1`] = `
"
class a implements b<c> {}

      ↓ ↓ ↓ ↓ ↓ ↓

class a implements b<c> {}
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`43. function type return 1`] = `
"
type a = () => b

      ↓ ↓ ↓ ↓ ↓ ↓

type a = () => b;
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`44. function type param 1`] = `
"
type a = (b: c) => {}

      ↓ ↓ ↓ ↓ ↓ ↓

type a = (b: c) => {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`45. function type param unnamed 1`] = `
"
type a = (b) => {}

      ↓ ↓ ↓ ↓ ↓ ↓

type a = (: b) => {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
\`;
"
`;

exports[`46. function type type param 1`] = `
"
type a = <b>() => {}

      ↓ ↓ ↓ ↓ ↓ ↓

type a = <b>() => {};
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'binding'
      name: 'b'
\`;
"
`;

exports[`47. object type property 1`] = `
"
type a = { b: c }

      ↓ ↓ ↓ ↓ ↓ ↓

type a = { b: c };
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`48. object type method 1`] = `
"
type a = { b(): c }

      ↓ ↓ ↓ ↓ ↓ ↓

type a = { b: () => c };
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;

exports[`49. object type indexer 1`] = `
"
type a = { [b: c]: d }

      ↓ ↓ ↓ ↓ ↓ ↓

type a = { [b: c]: d };
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'static'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
  -
      kind: 'reference'
      name: 'd'
\`;
"
`;

exports[`50. object type computed property 1`] = `
"
type a = { [b]: c }

      ↓ ↓ ↓ ↓ ↓ ↓

type a = { [: b]: c };
\`
  -
      kind: 'binding'
      name: 'a'
  -
      kind: 'reference'
      name: 'b'
  -
      kind: 'reference'
      name: 'c'
\`;
"
`;
